### PR TITLE
SOC-5272 : Improve user suggestions when @mentioning in activities and invite member to space

### PR DESCRIPTION
**Pull request by thomasdelhomenie**

### CODE REVIEW

File: component/core/src/main/java/org/exoplatform/social/core/storage/cache/model/key/IdentityFilterKey.java [hunk 1]
```diff
@@ -67,6 +68,12 @@ public IdentityFilterKey(final String providerId, final String remoteId, final P
     }
 
     this.excluded = Collections.unmodifiableList(keys);
+
+    List<String> searchIn = new ArrayList<>();
+    if (filter.getSearchInIdentities() != null && filter.getSearchInIdentities().size() > 0) {
+      filter.getSearchInIdentities().forEach(s -> searchIn.add(s));
```
Comment (thomasdelhomenie) [inline component/core/src/main/java/org/exoplatform/social/core/storage/cache/model/key/IdentityFilterKey.java]: Unfortunately we cannot still use Java 8 since it must be compatible with PLF 4.3 which supports Java 7.

**Comment by thomasdelhomenie** [comment]
I did some test on acceptance. It works fine (connections are suggested first) but it must be tested on a bigger dataset to be validated correctly. We are working on setting up an environment with Tribe data.

Just noticed one issue, with accents. I had an user "Thomas Delhoménie", when I typed "Delhoménie" I saw my user in the suggestions but not when I typed "Delhomenie". Accents should be ignored.

File: component/core/src/main/java/org/exoplatform/social/core/storage/cache/model/key/IdentityFilterKey.java (hunk 1 above)
Comment (nttuyen) [inline component/core/src/main/java/org/exoplatform/social/core/storage/cache/model/key/IdentityFilterKey.java]: I see in develop branch and this feature branch, we have an enforce rule require build with Java 8.  So, Why can't use java 8 here, it's impossible to run on jdk7 if we build on jdk8

Comment (thomasdelhomenie) [inline component/core/src/main/java/org/exoplatform/social/core/storage/cache/model/key/IdentityFilterKey.java]: Yes, but we will then merge this PR in stable/4.3.x branch also, not only develop. So it has to be compatible with stable/4.3.x which builds with Java 7.
The other option is to do 2 different PRs for develop and stable, but I don't think it worths it in this case.

Comment (nttuyen) [inline component/core/src/main/java/org/exoplatform/social/core/storage/cache/model/key/IdentityFilterKey.java]: ok, I will update it.

**Comment by nttuyen** [comment]
Currently we are doing as what social project did before, I mean the accent characters are not ignored in searching, user has to remember exactly the name of people who he want to mentions.

After investigate we find that to support this function, we have to add 'org.exoplatform.services.jcr.analyzer.IgnoreAccentAnalyzer' into social workspace and use the contains condition instead of like condition in search api now (do like we did in wiki project - which ignore accent character when search).
But I'm not sure that we will have to re-index all data when we change the analyzer or not? And with this change the search api in social may return more result than what we did, so I'm not sure is it impact to any functionality or not too.

One more thing I want to notice is there might be a potential performance issue in this PR because in case of search user who has connected with current user and match with keyword we have to fetch all users (who connected with) then compare (with keyword) manually. It might take time if user have many connections and none of them match with keyword.
We can not use JCR query for searching in this case because JCR does not support the join query currently.

**Comment by thomasdelhomenie** [comment]
OK, let's focus on the new mention algorithm and split the problem with accents in another issue. I will create a new Jira issue.

> One more thing I want to notice is there might be a potential performance issue in this PR because in case of search user who has connected with current user and match with keyword we have to fetch all users (who connected with) then compare (with keyword) manually. It might take time if user have many connections and none of them match with keyword.

Regarding his number of connections, I guess Patrice will not like that :)
Can you describe more in details how the mention search worked before and after your PR please ? I guess that before it was only one JCR search in all Social profiles (which are indexed), right ?

**Comment by nttuyen** [comment]
You are right, before it was only one JCR search in all social profile. But after apply this PR, we will have to do some more search by these steps:

- Firstly, If in case of activity, we have to check that creator is matched or not to add to result
- Then we will search in users who commented this activity if has
- Then we will search in users who liked this activity if has
- Then we will search in users who mentioned in if has

- Then if in space context, we will search in users belong to this space
- Then we will search in users who connected to current user (this case we have to search manually as I described)
- Then we will search with JCR query in all users (like what we did before)

=> To implement the mention improvement, we have to make query from top to bottom until we got enough results. So, In the worst case, we have to make 7 search call => There might be a performance issue here.

**Comment by thomasdelhomenie** [comment]
Indeed, looks like it will have impacts on the performances... We have to test it carefully.
Is it possible to send the suggested users after each step, to start proposing suggestions as soon as possible ?

Also we should think of this use case for social-rdbms. I guess Elasticsearch can help here if we index things correctly.

**Comment by nttuyen** [comment]
> Is it possible to send the suggested users after each step, to start proposing suggestions as soon as possible ?
 
I think it's impossible now, rest framework does not support to send response part by part and stop the process when user got his expected

> we should think of this use case for social-rdbms. I guess Elasticsearch can help here if we index things correctly.

Yes, with ES we can use [function score query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-function-score-query.html) to increase the score of users who connected with current user (or are member of space,...) and make these users at the top of result. To do this, we will need to index the connections (we did it in social-rdbms addon) and all space of user.

Here is my test with ES to query users match with keyword and make the connected user into the first of result:
```
GET /profile/_search
{
  "query": {
    "function_score": {
      "query": {
        "bool": {
          "should": [
            {
              "query_string": {
                "query": "*u*"
              }
            }
          ]
        }
      },
      "functions": [
        {
          "filter": {
            "bool": {
              "should": [
                {
                  "query": {
                    "query_string": {
                      "query": "2",
                      "fields": ["_id"]
                    }
                  }
                }
              ]
            }
          },
          "weight": 3
        },
        {
          "filter": {
            "bool": {
              "should": [
                {
                  "query": {
                    "query_string": {
                      "query": "*1*",
                      "fields": ["connections"]
                    }
                  }
                }
              ]
            }
          }, 
          "weight": 2
        }
      ]
    }
  }
}
```

**Comment by thomasdelhomenie** [comment]
> I think it's impossible now, rest framework does not support to send response part by part and stop the process when user got his expected
OK, let's wait for the first tests and see. No premature optimization.

About ES, we have to be able to switch easily from the current implementation to a new implementation using ES. Does it imply changes in the current code of social or social-rdbms ?

**Comment by nttuyen** [comment]
> About ES, we have to be able to switch easily from the current implementation to a new implementation using ES. Does it imply changes in the current code of social or social-rdbms ?

We need to change in social-rdbms only.

**Comment by thomasdelhomenie** [comment]
> We need to change in social-rdbms only.

:+1:

**Comment by boubaker** [comment]
Not used anymore...
replaced by:
https://github.com/exodev/social/pull/176
